### PR TITLE
Do not cause a stack overflow if logging fails

### PR DESCRIPTION
**Pull request by eric-wieser**
See individual commit messages. Needed to debug what's going wrong in #1313, after #1320 wasn't enough.

### CODE REVIEW

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 1]
```diff
@@ -99,6 +99,41 @@ const char *log_level(long level)
 #define LOG_SIZE    512
 static char log_buff[LOG_SIZE];
 
+
+/** Log as a backup, if python isn't working */
+static void gpi_log_native_v(const char *name, long level, const char *pathname, const char *funcname, long lineno, const char *msg, va_list argp) {
```
Comment (imphil) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Can you please wrap that line and place the `{` on a separate line for consistent style within this file?

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 2]
```diff
@@ -99,6 +99,41 @@ const char *log_level(long level)
...
+
+        fprintf(stdout, "%-4ld", lineno);
+        fprintf(stdout, " in %-31s ", funcname);
+        fprintf(stdout, "%s", log_buff);
+        fprintf(stdout, "\n");
+        fflush(stdout);
+    }
+}
+
+static void gpi_log_native(const char *name, long level, const char *pathname, const char *funcname, long lineno, const char *msg, ...) {
```
Comment (imphil) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: as above: please wrap and put `{` on a separate line.

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 3]
```diff
@@ -99,6 +99,41 @@ const char *log_level(long level)
 #define LOG_SIZE    512
 static char log_buff[LOG_SIZE];
 
+
+/** Log as a backup, if python isn't working */
```
Comment (imphil) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Can you be more descriptive when this function will be used? What conditions make Python not working?

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 4]
```diff
@@ -99,6 +99,41 @@ const char *log_level(long level)
...
+static void gpi_log_native_v(const char *name, long level, const char *pathname, const char *funcname, long lineno, const char *msg, va_list argp) {
+    if (level >= GPIInfo) {
```
Comment (imphil) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Please invert the condition and return early to reduce the indentation of this block.

```c
if (level < GPIInfo) {
    return;
}
```

Comment (eric-wieser) [reply cocotb/share/lib/gpi_log/gpi_logging.c]: This was done this way to match what was here before - I'll add an extra commit with this change.

**Comment by eric-wieser** [comment]
Comments resolved I think

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 5]
```diff
@@ -99,6 +99,50 @@ const char *log_level(long level)
...
+    int n = vsnprintf(log_buff, LOG_SIZE, msg, argp);
+    if (n < 0) {
+        fprintf(stderr, "Log message construction failed\n");
+    }
+
+    fprintf(stdout, "     -.--ns ");
+    fprintf(stdout, "%-9s", log_level(level));
+    fprintf(stdout, "%-35s", name);
+
+    n = strlen(pathname);
```
Comment (ktbarrett) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Reusing variables is bad practice.

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 6]
```diff
@@ -187,9 +202,8 @@ void gpi_log(const char *name, long level, const char *pathname, const char *fun
     }
 
     // Ignore truncation
-    va_start(ap, msg);
-    n = vsnprintf(log_buff, LOG_SIZE, msg, ap);
-    va_end(ap);
+    int n = vsnprintf(log_buff, LOG_SIZE, msg, argp);
+    (void) n;
```
Comment (ktbarrett) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: we have `COCOTB_UNUSED` to mark unused variables.

File: cocotb/share/lib/gpi_log/gpi_logging.c (hunk 5 above)
Comment (ktbarrett) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: Reusing variables is bad practice.

File: cocotb/share/lib/gpi_log/gpi_logging.c (hunk 6 above)
Comment (ktbarrett) [reply cocotb/share/lib/gpi_log/gpi_logging.c]: nvm, that hasn't been merged yet. #1137.

File: cocotb/share/lib/gpi_log/gpi_logging.c (hunk 5 above)
Comment (eric-wieser) [reply cocotb/share/lib/gpi_log/gpi_logging.c]: Perhaps, but this is just old code that moved.

**Comment by imphil** [comment]
I just merged #1137, which now requires a slight rework of this PR unfortunately. (I thought it would be easier to rework this one than #1137, which also has been open for quite some time. Sorry @eric-wieser.)

**Comment by eric-wieser** [comment]
Rebased, hopefully correctly

File: cocotb/share/lib/gpi_log/gpi_logging.c [hunk 7]
```diff
@@ -186,14 +203,12 @@ void gpi_log(const char *name, enum gpi_log_levels level, const char *pathname,
     }
 
     // Ignore truncation
-    va_start(ap, msg);
     {
         int n = vsnprintf(log_buff, LOG_SIZE, msg, ap);
```
Comment (eric-wieser) [inline cocotb/share/lib/gpi_log/gpi_logging.c]: argp, whoops

**Comment by eric-wieser** [comment]
CI failure is some readthedocs pip issue, nothing to do with this patch.

**Comment by eric-wieser** [comment]
@imphil, mind a quick re-review to check comments are addressed?

**Pull request merged**